<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Detection of Gene Expression Mid-parent Heterosis • TwoStageLRT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Detection of Gene Expression Mid-parent Heterosis">
<meta property="og:description" content="A two stage likelihood ratio test for detection of gene expression mid-parent heterosis from unreplicated multi-family RNA-seq experiments.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">TwoStageLRT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/TwoStageLRT.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/yunhuiqistat/TwoStageLRT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="twostagelrt">TwoStageLRT<a class="anchor" aria-label="anchor" href="#twostagelrt"></a>
</h1></div>
<!-- badges: start -->

<p>The goal of TwoStageLRT (2sLRT) is to detect gene expression mid-parent heterosis (MPH) from unreplicated multi-family RNA-seq experiment.</p>
<p>If you want to use 2sLRT, you might have gene expression data for multiple genes and multiple families, each family has three genotypes, female parent, male parent and their hybrid genotype. The key feature of using 2sLRT is there is only one sample for each genotype, i.e., unreplicated.</p>
<p>This test has two stages to deal with the problem of lacking replications. In the first stage, dispersion paremters are estimated. In the second stage, likelihood ratio test based on negative binomial distribution given the estimated dispersion parameters is conducted.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of TwoStageLRT like so:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"YunhuiQi/TwoStageLRT"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="website">Website<a class="anchor" aria-label="anchor" href="#website"></a>
</h2>
<p>Check the vignette at <a href="https://yunhuiqi.github.io/TwoStageLRT/" class="external-link uri">https://yunhuiqi.github.io/TwoStageLRT/</a> .</p>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p>To use 2sLRT, you need to have at least three data matrices:</p>
<ul>
<li><p>female_mat: female expression count with genes in rows, samples in columns.</p></li>
<li><p>male_mat: male expression count with genes in rows, samples in columns.</p></li>
<li><p>hybrid_mat: hybrids expression count with genes in rows, samples in columns.</p></li>
</ul>
<p>You can also specify the dispersion matrix, normalization factors if you have prior knowledge.</p>
<ul>
<li><p>disp_mat if “Pois”, use Poisson LRT and only output Pois results; if NULL, use dispersion estimation output will have both NBLRT and PoisLRT results; else, use provided disp_mat for NB test, output will have both NBLRT and PoisLRT results.</p></li>
<li><p>normalization logical, whether or not need normalization for the raw data matrices, if FALSE, treat the normalization factors all 1.</p></li>
<li><p>norm_factors if normalization = TRUE and norm_factors = NULL, use DESeq normalization factors; if normalization = TRUE and norm_factors is a list contains three vectors female, male and hybrid normalization factors, use the provided norm_factors in tests.</p></li>
</ul>
<p>Besides, you should also specify the hyper-parameters for dispersion estimation:</p>
<ul>
<li><p>n_fam_thres genes with at least n_fam_thres null families will be used for dispersion estimation, otherwise, will use loess prediction to get dispersion, default is 20.</p></li>
<li><p>gene_groups_no the number of clusters for genes, default is 3.</p></li>
</ul>
<p>2sLRT will give results based on your choice:</p>
<ul>
<li><p>if disp_mat = “Pois”, return Poisson LRT for heterosis pvalue and adjusted pvalue from BH adjustment;</p></li>
<li><p>if disp_mat = NULL or provided, return both NB and Poisson LRT for heterosis pvalue and qvalue and genewise dispersion, and the estimation of mu_gi, which is the female/male/hybrid_mat_normed.</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yunhuiqistat/TwoStageLRT" class="external-link">TwoStageLRT</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sim_data</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/twoStageLRT.html">twoStageLRT</a></span><span class="op">(</span>female_mat <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">female_mat</span>, male_mat <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">male_mat</span>,</span>
<span>                   hybrid_mat <span class="op">=</span> <span class="va">sim_data</span><span class="op">$</span><span class="va">hybrid_mat</span>, disp_mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                   normalization <span class="op">=</span> <span class="cn">TRUE</span>, norm_factors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                   n_fam_thres <span class="op">=</span> <span class="fl">10</span>, gene_groups_no <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Use DESeq estimated normalization factors. </span></span>
<span><span class="co">#&gt; Gene clusters size: </span></span>
<span><span class="co">#&gt; jacc_clust</span></span>
<span><span class="co">#&gt;  1  2 </span></span>
<span><span class="co">#&gt; 38  6 </span></span>
<span><span class="co">#&gt; Dispersion estimation for cluster 1 the number of families is 10 , number of genes is  38  </span></span>
<span><span class="co">#&gt; Dispersion estimation for cluster 2 the number of families is 14 , number of genes is  6  </span></span>
<span><span class="co">#&gt; Dispersion estimation summary:</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's </span></span>
<span><span class="co">#&gt; 0.02848 0.07723 0.09988 0.09786 0.11958 0.20517      56 </span></span>
<span><span class="co">#&gt; Output is from NB LRT for heterosis.</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pval_NB</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">#&gt; 0.000000 0.006074 0.140645 0.287458 0.537483 0.999942</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pval_Pois</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">#&gt; 0.000000 0.000000 0.000000 0.067466 0.002527 0.999684</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">DISP</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt; 0.02848 0.08031 0.09892 0.09567 0.11058 0.20517</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/yunhuiqistat/TwoStageLRT/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/yunhuiqistat/TwoStageLRT/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing TwoStageLRT</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yunhui Qi <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://opensource.org/licenses/MIT" class="external-link"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yunhui Qi.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
